### ISSUE TITLE
Browser stack overflow while serializing Suite object in WebDriver reporter

### ISSUE THREAD

**Issue by croes** [theintern/intern#525]
When trying to upgrade from Intern 2.2.2 to Intern 3.0.6, I've encountered the following issue.
All tests seems to run fine, but at the end of the run, the run crashes:

The following errors are reported:

```
Reporter error!
InternalError: too much recursion
  at Suite.prototype.toJSON  <__intern/lib/Suite.js:417:5>
  at sendRequest  <__intern/lib/reporters/WebDriver.js:146:12>
  at WebDriver.prototype._send  <__intern/lib/reporters/WebDriver.js:176:5>
  at WebDriver.prototype._sendEvent  <__intern/lib/reporters/WebDriver.js:130:11>
  at WebDriver.prototype.suiteEnd  <__intern/lib/reporters/WebDriver.js:54:11>
  at ReporterManager.prototype.emit/<  <__intern/lib/ReporterManager.js:250:20>
  at ReporterManager.prototype.emit  <__intern/lib/ReporterManager.js:232:23>
  at report  <__intern/lib/Suite.js:199:13>
  at end  <__intern/lib/Suite.js:193:12>
  at Promise/this.then/</<  <__intern/node_modules/dojo/Promise.js:156:41>
  at runCallbacks  <__intern/node_modules/dojo/Promise.js:19:13>
  at settle/<  <__intern/node_modules/dojo/Promise.js:103:21>
  at run  <__intern/node_modules/dojo/Promise.js:51:33>
  at CallbackQueue</CallbackQueue.prototype.drain  <__intern/node_modules/dojo/CallbackQueue.js:39:21>
  at nextTick</observer<  <__intern/node_modules/dojo/nextTick.js:52:21>

Error: One or more suite errors occurred during testing
    at http://localhost:8072/__intern/lib/executors/Executor.js:301:13
    at http://localhost:8072/__intern/node_modules/dojo/Promise.js:156:41
    at runCallbacks (http://localhost:8072/__intern/node_modules/dojo/Promise.js:19:22)
    at http://localhost:8072/__intern/node_modules/dojo/Promise.js:103:21
    at run (http://localhost:8072/__intern/node_modules/dojo/Promise.js:51:33)
    at CallbackQueue.drain (http://localhost:8072/__intern/node_modules/dojo/CallbackQueue.js:39:35)
    at MutationObserver.<anonymous> (http://localhost:8072/__intern/node_modules/dojo/nextTick.js:52:27)
```

Right before the crash, the Selenium server reports (not sure if this is relevant):

```
11:19:06.610 INFO - Executing: [execute script: return (function getCoverageData(coverageVariable) {
                var coverageData = (function () { return this; })()[coverageVariable];
                return coverageData && JSON.stringify(coverageData);
        }).apply(this, arguments);, [__internCoverage]])
11:19:06.717 INFO - Done: [execute script: return (function getCoverageData(coverageVariable) {
                var coverageData = (function () { return this; })()[coverageVariable];
                return coverageData && JSON.stringify(coverageData);
        }).apply(this, arguments);, [__internCoverage]]
```

The first error reported originates from a Firefox browser in this case, but I've seen similar 'stack overflow' errors for Chrome as well, except that it reports `RangeError: maximum call stack size exceeded` instead of `InternalError: too much recursion`.

Intern config:

``` javascript
define(function() {

   /*
    ...
    getProxyServerHostname()
    getProxyServerPort()
    getWebDriverHostname()
    ...
   */

  return {
    // The port on which the instrumenting proxy will listen
    proxyPort: 9000,

    // A fully qualified URL to the Intern proxy
    proxyUrl: 'http://' + getProxyServerHostname() + ':' + getProxyServerPort() + '/',

    // Default desired capabilities for all environments. Individual capabilities can be overridden by any of the
    // specified browser environments in the `environments` array below as well. See
    // https://code.google.com/p/selenium/wiki/DesiredCapabilities for standard Selenium capabilities and
    // https://saucelabs.com/docs/additional-config#desired-capabilities for Sauce Labs capabilities.
    // Note that the `build` capability will be filled in with the current commit ID from the Travis CI environment
    // automatically
    capabilities: {
    },

    // Browsers to run integration testing against. Note that version numbers must be strings if used with Sauce
    // OnDemand. Options that will be permutated are browserName, version, platform, and platformVersion; any other
    // capabilities options specified for an environment will be copied as-is
    environments: [
      { browserName: "firefox", version: "", platform: [ "LINUX" ] },
      { browserName: "firefox", version: "", platform: [ "WINDOWS" ] },
      { browserName: "chrome", version: "", platform: ["LINUX"]},
      { browserName: "chrome", version: "", platform: ["WINDOWS"] }
    ],

    // Maximum number of simultaneous integration tests that should be executed on the remote WebDriver service
    maxConcurrency: Number.POSITIVE_INFINITY,

    // Whether or not to start Sauce Connect before running tests
    tunnel: 'NullTunnel',

    // Connection information for the remote WebDriver service. If using Sauce Labs, keep your username and password
    // in the SAUCE_USERNAME and SAUCE_ACCESS_KEY environment variables unless you are sure you will NEVER be
    // publishing this configuration file somewhere
    tunnelOptions: {
      hostname: getWebDriverHostname(),
      port: 4444
    },

    // Non-functional test suite(s) to run in each browser
    suites: [
      "tests/all",
      "tests/symbology/all",
      "tests/geometry/constructive/all"
    ],

    reporters: [
      { id: 'Runner' },
      { id: "ria/test/javascript/reporters/XunitReporter3" }
    ],

    // Functional test suite(s) to run in each browser once non-functional tests are completed
    functionalSuites: [],

    // A regular expression matching URLs to files that should not be included in code coverage analysis
    excludeInstrumentation: /^(?:ria\/node_modules|ria\/test)\//
  };
});
```

Some environment details:
- Intern 3.0.6
- Selenium Grid 2.46.0 (also managed to reproduce it on a local installation with version 2.48.2)
- The entire test run contains about 5k tests

The tests are being run on a Selenium Grid with these environments:
- Firefox Windows (39.0)
- Firefox Linux (39.0)
- Chrome Linux (46.0)
- Chrome Windows (46.0)

Running the tests with the Intern browser client (on 3.0.6) does not cause the stack overflow error, the test run finishes normally.
The issue also does not occur when running the same tests with Intern 2.2.2.

Any ideas why the browser's stack overflows while (apparently) serializing the Suite object in the WebDriver reporter?

**Comment by croes** [comment]
Upon further investigation I found that there is a `Suite` object that has an error attached to it. `suite.error.relatedTest` points to the Suite itself, causing infinite recursion when trying to serialize the object with JSON.stringify.

I will investigate further exactly how the Suite object ends up pointing to itself in it's `error.relatedTest`.

**Comment by croes** [comment]
Some more remarks:

* The issues are caused by some `Suite`s throwing errors in `after`/`teardown`
* `suite.error.relatedTest === suite` if the `teardown` / `after` function of the Suite throws an error. This gives issues when serializing the `Suite` object in the `WebDriver` reporter. I solved this by not serializing the `relatedTest` property of `error.relatedTest` if it equals the suite itself (`this`).
* If the `after` of the final suite throws, the error ends up at the `PreExecutor`, which will send a `fatalError` message to the Node Runner. The sequence number of this message is 0, while the `WebDriver` reporter has already sent messages with larger sequence numbers. The `Proxy` reports the following error:

```
Error: Repeated sequence for session 3864e420-dcff-4b06-bd21-7e74e9786f3b: 18 last 0 cur
  at Proxy._publishInSequence  <ria/node_modules/intern/lib/Proxy.js:185:11>
  at IncomingMessage.<anonymous>  <ria/node_modules/intern/lib/Proxy.js:67:39>
  at IncomingMessage.EventEmitter.emit  <events.js:92:17>
  at <_stream_readable.js:920:16>
  at process._tickCallback  <node.js:415:13>
```

I solved this by not checking the sequence number for these 'special' fatalError messages (see below) This allows the Node runner process to finish gracefully. Another solution would be to somehow send the correct sequence number.

Proxy.js
```javascript
function isFatalErrorFromPreExecutor(message) {
  message.payload[0] === 'fatalError' && message.payload[1].sequence === 0;
}

//...

if (!isFatalErrorFromPreExecutor(message) && message.sequence <= session.lastSequence) {
	throw new Error('Repeated sequence for session ' + message.sessionId + ': ' + session.lastSequence +
					' last ' + message.sequence + ' cur');
}
```

**Comment by jason0x43** [comment]
This issue was resolved with #561.
